internal/pkg/api/server/templates/dependencies/locks.go

```diff
@@ -11,68 +12,84 @@ import (
-	return &Locker{d: d, ttl: ttl}
+func NewLocker(d lockerDeps, ttlSeconds int) *Locker {
+	return &Locker{d: d, ttlSeconds: ttlSeconds}
 }
 
-func (l *Locker) Lock(requestCtx context.Context, lockName string) (bool, UnlockFn) {
-	// Get client
-	c, err := l.d.EtcdClient(requestCtx)
-	if err != nil {
-		l.d.Logger().Warnf(`cannot acquire etcd lock "%s" (continues without lock): cannot get etcd client: %s`, lockName, err.Error())
-		return true, func() {}
-	}
-
-	// Get concurrency session with TTL
-	session, err := concurrency.NewSession(c, concurrency.WithTTL(l.ttl))
-	if err != nil {
-		l.d.Logger().Warnf(`cannot acquire etcd lock "%s" (continues without lock): cannot obtain session: %s`, err.Error())
-		return true, func() {}
-	}
-
-	// Context with acquire timeout
-	acquireCtx, cancelFn := context.WithTimeout(requestCtx, LockAcquireTimeout)
-	defer cancelFn()
-
+func (l *Locker) TryLock(requestCtx context.Context, lockName string) (bool, UnlockFn) {
 	// Try lock
-	mtx := concurrency.NewMutex(session, lockName)
-	if err := mtx.TryLock(acquireCtx); errors.Is(err, concurrency.ErrLocked) {
+	session, mtx, err := l.tryLock(requestCtx, lockName)
+	if errors.Is(err, concurrency.ErrLocked) {
 		l.d.Logger().Infof(`etcd lock "%s" is used`, lockName)
 		return false, func() {}
 	} else if err != nil {
-		l.d.Logger().Warnf("cannot acquire etcd lock (continues without lock): cannot lock mutex: %s", err.Error())
+		l.d.Logger().Warnf(`cannot acquire etcd lock "%s" (continues without lock): %s`, lockName, err)
 		return true, func() {}
 	}
 
-	// End the refresh for the session lease, so TTL will apply
-	session.Orphan()
-
 	// Locked, must be unlocked by returned unlock function
 	l.d.Logger().Infof(`acquired etcd lock "%s"`, mtx.Key())
 	return true, func() {
 		releaseCtx, cancelFn := context.WithTimeout(context.Background(), LockReleaseTimeout)
 		defer cancelFn()
 		if err := mtx.Unlock(releaseCtx); err != nil {
-			l.d.Logger().Warnf(`cannot unlock etcd lock "%s": %s`, mtx.Key(), err.Error())
+			l.d.Logger().Warnf(`cannot unlock etcd lock "%s": %s`, lockName, err.Error())
 		}
 		if err := session.Close(); err != nil {
-			l.d.Logger().Warnf(`cannot close etcd session for lock "%s": %s`, mtx.Key(), err.Error())
+			l.d.Logger().Warnf(`cannot close etcd session for lock "%s": %s`, lockName, err.Error())
 		}
 		l.d.Logger().Infof(`released etcd lock "%s"`, lockName)
 	}
 }
+
+func (l *Locker) tryLock(requestCtx context.Context, lockName string) (*concurrency.Session, *concurrency.Mutex, error) {
+	// Get client
+	c, err := l.d.EtcdClient(requestCtx)
+	if err != nil {
+		return nil, nil, fmt.Errorf(`cannot get etcd client: %w`, err)
+	}
+
+	// Acquire timeout
+	acquireCtx, cancelFn := context.WithTimeout(requestCtx, LockAcquireTimeout)
+	defer cancelFn()
+
+	// Creates a new lease
+	lease, err := c.Grant(acquireCtx, int64(l.ttlSeconds))
+	if err != nil {
+		return nil, nil, fmt.Errorf(`cannot grant lease %w`, err)
+	}
```

Comment:
Tu pribudol tento blok.
- `lease` je nieco co ma `TTL - time to live`, ktory plati pre KV hodnoty aj pre zamky (my pouzivame iba zamky).
- metoda `concurrency.NewSession` pouzita nizsie si vie `lease` vytvorit aj sama, takto sa to robilo predtym, vsimni si TTL: https://github.com/keboola/keboola-as-code/blob/97c7d1641d3a7b16b1361588e8d74c53a3a8e730/internal/pkg/api/server/templates/dependencies/locks.go#L41-L46
- no ako mi nefungovalo to lokalne `etcd`, tak som si vsimol, ze nenastane timeout po `2 sekundach`
- je tam totiz problem s `context`:
   - session ma trvat pocas celeho requestu
   - ale `grant lease` maximalne `2 sekundy`.
- preto som to musel takto oddelit, aby som ^^^ zabezpecil.
- testoval som to manualne a bez etcd to pekne po `2 sekundach` skoncilo
- TLDR: `grant lease` ma `acquireCtx`, `NewSession` ma `requestCtx`.